### PR TITLE
ship LLVM tools with the toolchain

### PR DESCRIPTION
**Pull request by japaric**
this PR adds llvm-{nm,objcopy,objdump,size} to the rustc sysroot (right next to LLD)

this slightly increases the size of the rustc component. I measured these numbers on x86_64 Linux:

- rustc-1.27.0-dev-x86_64-unknown-linux-gnu.tar.gz 180M -> 193M (+7%)
- rustc-1.27.0-dev-x86_64-unknown-linux-gnu.tar.xz 129M -> 137M (+6%)

r? @alexcrichton 
cc #49584

### CODE REVIEW

**Comment by emilio** [comment]
Could you add `llvm-profdata` too if it's not too much to ask? It's useful / necessary for PGO.

Otherwise I can do it in a followup myself I guess. Thanks for working on this!

**Comment by japaric** [comment]
@emilio added llvm-profdata

**Comment by alexcrichton** [comment]
@bors: try

I'm curious to poke around these from the bots as well

File: src/bootstrap/lib.rs [hunk 1]
```diff
@@ -199,6 +199,10 @@ use flags::Subcommand;
 use cache::{Interned, INTERNER};
 use toolstate::ToolState;
 
+const LLVM_TOOLS: &[&str] = &[
+    "llvm-nm", "llvm-objcopy", "llvm-objdump", "llvm-profdata", "llvm-size",
```
Comment (alexcrichton) [inline src/bootstrap/lib.rs]: Could you add some comments here for what each tool is and why it's included?

**Comment by alexcrichton** [comment]
The code changes here look good to me, thanks @japaric! I'd like to poke around the output and then I think we can FCP for inclusion with... some team (?)

**Comment by alexcrichton** [comment]
It looks like the compiled tools have a dynamic dependency on `libstdc++.so.6` which generally isn't suitable for shipping to end users (rustc links libstdc++ statically), so I think we'll need to figure out how to coerce LLVM's build system to link libstdc++ statically before merging this regardless.

**Comment by alexcrichton** [comment]
@Mark-Simulacrum a downside to making this a component, however, is that it's "yet another step" to get up and running with embedded development. This could perhaps be folded into an existing step if it already existed, but @japaric it sounds like there's not a great place for this to get folded into?

I previously hadn't been able to articulate a great downside to shipping this (and I agree that it's defacto stabilizing), but thinking now I think I may have come up with one (although seriousness of it TBD). This runs the risk of bifurcating "rust through rustup" and "rust through system installers". It seems likely that distros would delete these tools and/or would not ship them by default. That means that the experience in the embedded world would be that you'd have to deal with either the absence of the tools or perhaps a variety of versions (not great).

Now this may not actually matter too much. AFAIK the distro "cross compile" experience is definitely not as good as it is through rustup, so it may be perhaps a safe assumption that everyone using these tools are already using rustup and are cross compiling, meaning it's fine to ship/stabilize.

One aspect I particularly like about this is that we're not shipping these tools into users' PATH, but rather it's buried in a rustc-specific location that you have to go out of your way to access (which tools I believe @japaric is envisioning for embedded will do)

**Comment by cuviper** [comment]
> I think we'll need to figure out how to coerce LLVM's build system to link libstdc++ statically

[Fedora LLVM used to do this](https://src.fedoraproject.org/cgit/rpms/llvm.git/commit/?id=025cf6ec25c60f828b584d72f8ff9173fbd95bfe) (for Steam's sake) using:
`-DCMAKE_SHARED_LINKER_FLAGS="-Wl,-Bsymbolic -static-libstdc++"`

But I think you'll need `CMAKE_STATIC_LINKER_FLAGS` and/or `CMAKE_EXE_LINKER_FLAGS`.

> It seems likely that distros would delete these tools and/or would not ship them by default.

In Fedora they're built and shipped as part of the `llvm` package.  By default, `rustc` will only pull in the library in the `llvm-libs` subpackage though.  We also don't install `lld` by default.  In general, I would expect our system compilation with Rust to continue using `cc` (gcc) and binutils.

We don't generally support cross compilation at all though, so as far as I'm concerned, you can assume such users have rustup toolchains.  Just don't break my native compilation, please! 😄

**Comment by Mark-Simulacrum** [comment]
I don't really object to installing it by default; that way it's not a roadblock. I just don't really feel like it fits well into the `rustc` component. Having it as a separate component would also enable us to remove it with very visible messaging (instead of "file not found" messages) and/or ship it as a -preview component, for example.

If I understand correctly that these tools are primarily intended for embedded development, perhaps there's a way we can have them auto-install as part of the "cargo template" solution we're looking for? In theory there's no particular reason a build script or template script (if these exist?) couldn't go fetch them from static.r-l.o...

**Comment by japaric** [comment]
@cuviper Thanks for the hint. Setting CMAKE_EXE_LINKER_FLAGS seems to have done the trick.

I'm seeing 1MB increase in both the tar.gz and the tar.xz tarballs after statically linking to libstdc++ *and* adding llvm-profdata.

**Comment by alexcrichton** [comment]
@japaric is there perhaps a good location in the "embedded setup workflow" to automatically inject installation of these tools?

**Comment by alexcrichton** [comment]
Ok the binary artifacts themselves look good to me! I think we just need to figure out "component or not" now

**Comment by japaric** [comment]
@alexcrichton the setup looks like this:

- Install OpenOCD (a) and arm-none-eabi-gdb (b). Once and distro specific
- `rustup target add thumbv7m-none-eabi`. Once, per target.
- `cargo install cargo-binutils`. Once-ish (may need to be updated every now and then)
- Instantiate Cargo template. Per binary project.

(a) Distro specific installation until someone Rewrites It In Rust
(b) Until lldb ships with rust *and* we figure out how to make it work as nicely as GDB (i.e. OpenOCD integration)

We could do one of these to not increase the number of setup steps:

- Install the LLVM tools with the `rustup target add` command ... but there's nothing target specific about the tools; they are also useful to inspect native binaries (also profiling).
- Have the `cargo-binutils` subcommand implicitly install the `llvm-tools` rustup component (if not there already) when invoked ... but some people don't like automatic downloads.
- Have the Cargo template install the `llvm-tools` rustup component on instantiation ... but I don't know if that's even in scope for the Cargo template feature.

**Comment by jcsoo** [comment]
I'll make the argument that these tools belong with rustc because they are used for introspecting and manipulating the output of the compiler itself, and it's an oversight that they haven't been included in the past. `objcopy` could easily be implemented in rustc as an additional flag for specifying an output format, but for various reasons it's been split out into a separate binary, just like `lld`.

This is different than `rustfmt` or `rls`, tools primarily concerned with introspecting or manipulating user code.

If these tools were much bigger (tens or hundreds of megabytes), I would be sympathetic with the argument that we should create a new component for them, but I don't think it's a great user experience to have a component that we advertise to the user but that turns out in some cases to do nothing once they install it.

**Comment by alexcrichton** [comment]
Ok thanks for the info! With all that in mind I'm personally still tempted to ship these by default and starting to leverage the in the ecosystem.

I'd want to also cc @rust-lang/dev-tools, however, as this is likely most directly impacting y'all!

**Comment by Mark-Simulacrum** [comment]
With these arguments in mind I also agree that at least for now we can probably ship these in the rustc component and re-evaluate if necessary later down the line. I'd like for there to be documentation somewhere semi-official (e.g., the forge, maybe an RFC) that states the APIs of these tools and their existence is not part of Rust's stability promises. I imagine that in practice we'll do our best to avoid breakage but I would like to have that stated somewhere. (Of course, this is presuming we agree with such a guarantee.)

**Comment by michaelwoerister** [comment]
Would this help with https://github.com/rust-lang/rust/pull/50000, where I'd like `llvm-objdump` to be available for `run-make` tests?

**Comment by nrc** [comment]
We discussed this at the dev-tools meeting today - we all liked the general idea, but we did think that this ought to be a rustup component, unless there are big difficulties in doing so.

**Comment by TimNN** [comment]
@japaric / @alexcrichton: After the result from the dev-tools meeting, what is the way forward for this PR?

**Comment by alexcrichton** [comment]
I believe the next step here is to separate out the llvm tools into a separate rustup component

**Comment by japaric** [comment]
I'm swamped right now and won't have time to work on this until Rustfest impl days (at the earliest) so I'm going to close this to clear the PR queue. If someone pick this up before I get to it please go ahead!

**Comment by japaric** [comment]
I copy pasted the rustfmt component logic to make an llvm-tools component and bootstrap now produces a llvm-tools tarball that looks correct-ish but I'm not sure if my changes are enough to produce a correct manifest. `x.py dist llvm-tools` doesn't seem to produce a rustup manifest (the one that list all components).

File: src/bootstrap/lib.rs [hunk 2]
```diff
@@ -931,6 +939,24 @@ impl Build {
         self.package_vers(&self.release_num("rustfmt"))
     }
 
+    fn llvm_tools_vers(&self) -> String {
```
Comment (japaric) [inline src/bootstrap/lib.rs]: any suggestion to get the LLVM version from source code rather than from a binary?

**Comment by oli-obk** [comment]
@bors retry

sorry, that was me

**Comment by oli-obk** [comment]
@bors r-

oh that was travis, not bors

**Comment by japaric** [comment]
> Hm so actually for safety it's probably best to "version" these the same as rustc itself

Done

@bors r=alexcrichton

**Comment by TimNN** [comment]
@japaric: Looks like your PR failed on travis, do you think you'll be able to look into that?

**Comment by kennytm** [comment]
@bors r- 

Wrongly requeued.

**Comment by japaric** [comment]
Looking at how `-static-libstdc++` is used in `native::configure_cmake` I'm omitting it for Windows
like `configure_cmake` does. I'm also omitting `-Bsymbolic` on macOS as its C toolchain doesn't
support it (as evidenced in the Travis failure).

r? latest changes @alexcrichton

**Comment by kennytm** [comment]
@bors retry 

3 hour timeout?

File: src/bootstrap/lib.rs (hunk 2 above)
Comment (mati865) [reply src/bootstrap/lib.rs]: LLVM version is set when [running `cmake` on it's source code](https://github.com/llvm-mirror/llvm/blob/a71e59ebf7ff06a6c39d28d6fea4f9340c499d0f/CMakeLists.txt#L26).
After `cmake` was run the version is coded into [`llvm/Config/llvm-config.h`](https://github.com/llvm-mirror/llvm/blob/4833be099d271d1102c6d7c1fcd63d870a92a85f/include/llvm/Config/llvm-config.h.cmake#L75).

**Comment by vi** [comment]
Does it take into account theoretical future non-LLVM rustc?

**Comment by japaric** [comment]
@vi the tools are packaged in their own rustup component. rustc doesn't depend on these tools (or rustup). We are free to leave the llvm-tools component empty for some target platforms if we can't compile the tools for that platform (for any reason). We are also free to drop tools from the component if LLVM stops supporting them in the future.

I would be more concerned about lld which is always shipped with the toolchain and it's a hard requirement for cross compiling to wasm.

**Pull request merged**